src/dune_rules/stanzas/library.ml

```diff
@@ -118,7 +118,22 @@ let decode =
          "modes"
          (Modes.decode ~stanza_loc ~dune_version project)
          ~default:(Mode_conf.Lib.Set.default stanza_loc)
-     and+ kind = field "kind" Lib_kind.decode ~default:Lib_kind.Normal
+     and+ virtual_modules, kind =
+       let* virtual_modules =
+         Ordered_set_lang.Unexpanded.field_o
+           ~check:(Dune_lang.Syntax.since Stanza.syntax (1, 7))
+           ~since_expanded:Stanza_common.Modules_settings.since_expanded
+           "virtual_modules"
+       in
+       let+ kind = field "kind" Lib_kind.decode ~default:Lib_kind.Normal in
+       let kind : Lib_kind.t =
+         match virtual_modules, kind with
+         | Some _, (Normal | Virtual) -> Virtual
+         | None, kind -> kind
+         | Some _, _incompatible_kind ->
+           raise (Failure "TODO: Error on incompatible kind")
```

Comment:
What's the right way to signal an error here? `User_error.raise`?

Reply:
If you want this to be an error to the user then `User_error.raise`. If it is a programming error it should be a `Code_error.raise`.